src/Microsoft.OData.Core/WriterValidationUtils.cs

```diff
@@ -433,6 +512,52 @@ internal static void ValidateEntityReferenceLink(ODataEntityReferenceLink entity
             return navigationProperty;
         }
 
+        /// <summary>
+        /// Validates the input <see cref="IEdmStructuredType"/> meets the derived type constaints on the <see cref="ODataNestedResourceInfo"/>.
+        /// </summary>
+        /// <param name="resourceType">The input resource type.</param>
+        /// <param name="metadataType">The type from metadata.</param>
+        /// <param name="nestedResourceInfo">The nested resource info.</param>
+        /// <param name="derivedTypeConstraints">The derived type constraints on the nested resource.</param>
+        internal static void ValidateDerivedTypeConstraintOnNestedResourceInfo(IEdmStructuredType resourceType, IEdmStructuredType metadataType, ODataNestedResourceInfo nestedResourceInfo, IEnumerable<string> derivedTypeConstraints)
+        {
+            if (resourceType == null || metadataType == null || derivedTypeConstraints == null || resourceType == metadataType)
+            {
+                return;
+            }
+
+            string fullTypeName = resourceType.FullTypeName();
+            if (derivedTypeConstraints.Any(c => c == fullTypeName))
+            {
+                return;
+            }
+
+            throw new ODataException(Strings.WriterValidationUtils_PropertyValueTypeNotAllowedInDerivedTypeConstraint(fullTypeName, nestedResourceInfo.Name));
+        }
+
+        /// <summary>
+        /// Validates the input <see cref="IEdmStructuredType"/> meets the derived type constaints on the <see cref="IEdmNavigationSource"/>.
+        /// </summary>
+        /// <param name="resourceType">The input resource type.</param>
+        /// <param name="metadataType">The type from metadata.</param>
+        /// <param name="navigationSource">The navigation source.</param>
+        /// <param name="derivedTypeConstraints">The derived type constraints on the nested resource.</param>
+        internal static void ValidateDerivedTypeConstraintOnNavigationSource(IEdmStructuredType resourceType, IEdmStructuredType metadataType, IEdmNavigationSource navigationSource, IEnumerable<string> derivedTypeConstraints)

```

Comment:
>ValidateDerivedTypeConstraintOnNavigationSource [](start = 29, length = 47)

ValidateDerivedTypeConstraintOnNavigationSource and ValidateDerivedTypeConstraintOnNestedResourceInfo have the same logic, except that one uses the name of the navigation source in the exception and the other uses the name from the nestedresourceinfo.  Instead of two overloads, unless you think you'll need the navigationsource (or additional information out of the nestedResourceInfo) in the future, how about just having a single overload that takes a string property "navigationSourceName", and pass the appropriate string from the caller?

Reply:
#resolved.